File: algorithms/pid_controller/cpp/test_pid_controller.cpp
/**
 * test_pid_controller.cpp
 *
 * C++ test harness for the pid_controller algorithm.
 * Reads the same JSON test vectors used by the MATLAB test harness,
 * runs the generated C++ function, and validates outputs within tolerance.
 *
 * Also writes cpp_outputs.json for equivalence comparison with MATLAB.
 */

#include <gtest/gtest.h>
#include <nlohmann/json.hpp>
#include <filesystem>
#include <fstream>
#include <string>
#include <vector>
#include <cmath>

// Generated header from MATLAB Coder
#include "pid_controller.h"

using json = nlohmann::json;
namespace fs = std::filesystem;

#ifndef TEST_VECTORS_DIR
#error "TEST_VECTORS_DIR must be defined at compile time"
#endif

#ifndef OUTPUT_DIR
#define OUTPUT_DIR "."
#endif

// ---- Test case data structure ----

struct TestCase {
    std::string name;
    std::string description;

    // Inputs
    double error;
    double integral;
    double prev_error;
    double kp;
    double ki;
    double kd;
    double dt;

    // Expected outputs
    double expected_output;
    double expected_new_integral;
    double expected_new_prev_error;

    // Tolerance
    double abs_tolerance;
};

// For Google Test to print test case names
std::string TestCaseName(const ::testing::TestParamInfo<TestCase>& info) {
    return info.param.name;
}

// ---- Load test vectors from JSON ----

std::vector<TestCase> LoadTestVectors(const std::string& dir) {
    std::vector<TestCase> cases;

    for (const auto& entry : fs::directory_iterator(dir)) {
        if (entry.path().extension() != ".json") continue;
        if (entry.path().filename() == "schema.json") continue;

        std::ifstream f(entry.path());
        if (!f.is_open()) continue;

        json data = json::parse(f);

        // Global tolerance defaults
        double global_abs_tol = 1e-10;
        if (data.contains("global_tolerance") &&
            data["global_tolerance"].contains("absolute")) {
            global_abs_tol = data["global_tolerance"]["absolute"].get<double>();
        }

        for (const auto& tc : data["test_cases"]) {
            TestCase t;
            t.name = tc["name"].get<std::string>();
            t.description = tc.value("description", "");

            // Parse inputs
            t.error = tc["inputs"]["error"].get<double>();
            t.integral = tc["inputs"]["integral"].get<double>();
            t.prev_error = tc["inputs"]["prev_error"].get<double>();
            t.kp = tc["inputs"]["kp"].get<double>();
            t.ki = tc["inputs"]["ki"].get<double>();
            t.kd = tc["inputs"]["kd"].get<double>();
            t.dt = tc["inputs"]["dt"].get<double>();

            // Parse expected outputs
            t.expected_output = tc["expected_output"]["output"].get<double>();
            t.expected_new_integral = tc["expected_output"]["new_integral"].get<double>();
            t.expected_new_prev_error = tc["expected_output"]["new_prev_error"].get<double>();

            // Per-case tolerance overrides global
            t.abs_tolerance = global_abs_tol;
            if (tc.contains("tolerance") && tc["tolerance"].contains("absolute")) {
                t.abs_tolerance = tc["tolerance"]["absolute"].get<double>();
            }

            cases.push_back(t);
        }
    }

    return cases;
}

// ---- Parameterized test ----

class PidControllerTest : public ::testing::TestWithParam<TestCase> {};

TEST_P(PidControllerTest, MatchesExpectedOutput) {
    const auto& tc = GetParam();

    // Output variables
    double output = 0.0;
    double new_integral = 0.0;
    double new_prev_error = 0.0;

    // Call generated C++ function
    pid_controller::pid_controller(
        tc.error, tc.integral, tc.prev_error,
        tc.kp, tc.ki, tc.kd, tc.dt,
        &output, &new_integral, &new_prev_error);

    // Validate outputs
    EXPECT_NEAR(output, tc.expected_output, tc.abs_tolerance)
        << "Output mismatch in test case: " << tc.name;
    EXPECT_NEAR(new_integral, tc.expected_new_integral, tc.abs_tolerance)
        << "Integral mismatch in test case: " << tc.name;
    EXPECT_NEAR(new_prev_error, tc.expected_new_prev_error, tc.abs_tolerance)
        << "Prev error mismatch in test case: " << tc.name;
}

INSTANTIATE_TEST_SUITE_P(
    TestVectors,
    PidControllerTest,
    ::testing::ValuesIn(LoadTestVectors(TEST_VECTORS_DIR)),
    TestCaseName
);

// ---- Write outputs for equivalence comparison ----

class CppOutputWriter : public ::testing::Environment {
public:
    void TearDown() override {
        auto cases = LoadTestVectors(TEST_VECTORS_DIR);
        json outputs = json::array();

        for (const auto& tc : cases) {
            double output = 0.0;
            double new_integral = 0.0;
            double new_prev_error = 0.0;

            pid_controller::pid_controller(
                tc.error, tc.integral, tc.prev_error,
                tc.kp, tc.ki, tc.kd, tc.dt,
                &output, &new_integral, &new_prev_error);

            json result;
            result["test_name"] = tc.name;
            result["actual_output"] = output;
            result["actual_new_integral"] = new_integral;
            result["actual_new_prev_error"] = new_prev_error;
            result["tolerance"] = tc.abs_tolerance;
            outputs.push_back(result);
        }

        // Write to output directory
        fs::path output_dir(OUTPUT_DIR);
        fs::create_directories(output_dir);
        std::ofstream f(output_dir / "cpp_outputs.json");
        f << outputs.dump(2);
    }
};

// Register the output writer
testing::Environment* const output_env =
    testing::AddGlobalTestEnvironment(new CppOutputWriter);

File: algorithms/pid_controller/matlab/test_pid_controller.m
function results = test_pid_controller(vectors_dir, output_dir)
%TEST_PID_CONTROLLER Validate pid_controller against JSON test vectors.
%
%   results = test_pid_controller(vectors_dir, output_dir)

    results.Passed = 0;
    results.Failed = 0;
    results.Total = 0;
    results.Details = {};

    vector_files = dir(fullfile(vectors_dir, '*.json'));

    for i = 1:length(vector_files)
        if contains(vector_files(i).name, 'schema')
            continue;
        end

        filepath = fullfile(vectors_dir, vector_files(i).name);
        fid = fopen(filepath, 'r');
        raw = fread(fid, inf, 'char');
        fclose(fid);
        test_data = jsondecode(char(raw'));

        global_abs_tol = 1e-10;
        if isfield(test_data, 'global_tolerance')
            if isfield(test_data.global_tolerance, 'absolute')
                global_abs_tol = test_data.global_tolerance.absolute;
            end
        end

        for t = 1:length(test_data.test_cases)
            tc = test_data.test_cases(t);
            results.Total = results.Total + 1;

            err = tc.inputs.error;
            integral = tc.inputs.integral;
            prev_error = tc.inputs.prev_error;
            kp = tc.inputs.kp;
            ki = tc.inputs.ki;
            kd = tc.inputs.kd;
            dt = tc.inputs.dt;

            [actual_output, actual_new_integral, actual_new_prev_error] = ...
                pid_controller(err, integral, prev_error, kp, ki, kd, dt);

            expected_output = tc.expected_output.output;
            expected_new_integral = tc.expected_output.new_integral;
            expected_new_prev_error = tc.expected_output.new_prev_error;

            if isfield(tc, 'tolerance') && isfield(tc.tolerance, 'absolute')
                abs_tol = tc.tolerance.absolute;
            else
                abs_tol = global_abs_tol;
            end

            ok = abs(actual_output - expected_output) <= abs_tol && ...
                 abs(actual_new_integral - expected_new_integral) <= abs_tol && ...
                 abs(actual_new_prev_error - expected_new_prev_error) <= abs_tol;

            if ok
                results.Passed = results.Passed + 1;
                fprintf('  PASS: %s\n', tc.name);
            else
                results.Failed = results.Failed + 1;
                fprintf('  FAIL: %s\n', tc.name);
            end

            detail.test_name = tc.name;
            detail.actual_output = actual_output;
            detail.actual_new_integral = actual_new_integral;
            detail.actual_new_prev_error = actual_new_prev_error;
            detail.tolerance = abs_tol;
            detail.passed = ok;
            results.Details{end+1} = detail;
        end
    end

    fprintf('\nResults: %d passed, %d failed, %d total\n', ...
        results.Passed, results.Failed, results.Total);

    if nargin >= 2 && ~isempty(output_dir)
        if ~exist(output_dir, 'dir')
            mkdir(output_dir);
        end
        output_path = fullfile(output_dir, 'matlab_outputs.json');
        fid = fopen(output_path, 'w');
        fprintf(fid, '%s', jsonencode(results.Details));
        fclose(fid);
        fprintf('MATLAB outputs written to: %s\n', output_path);
    end
end
